runtime/src/lib.rs

```diff
@@ -871,6 +871,78 @@ pub const BABE_GENESIS_EPOCH_CONFIG: sp_consensus_babe::BabeEpochConfiguration =
         allowed_slots: sp_consensus_babe::AllowedSlots::PrimaryAndSecondaryVRFSlots,
     };
 
+mod migrations {
+    #[allow(unused_imports)]
+    use super::*;
+
+    pub mod migrate_staking_to_bags_list {
+        use super::*;
+        use frame_election_provider_support::SortedListProvider;
+
+        pub struct MigrateStakingToBagsList;
+
+        impl frame_support::traits::OnRuntimeUpgrade for MigrateStakingToBagsList {
+            #[cfg(feature = "try-runtime")]
+            fn pre_upgrade() -> Result<sp_std::vec::Vec<u8>, sp_runtime::TryRuntimeError> {
+                use codec::Encode;
+
+                frame_support::ensure!(
+                    crate::System::last_runtime_upgrade_spec_version() == 5_002,
+                    "must upgrade linearly"
+                );
+
+                let to_migrate = pallet_staking::Validators::<Runtime>::count()
+                    + pallet_staking::Nominators::<Runtime>::count();
+                log::info!("👜 staking bags-list migration passes PRE migrate checks ✅");
+                Ok(to_migrate.encode())
+            }
+
+            /// Migrates validators and nominators to bags list for the staking pallet.
+            fn on_runtime_upgrade() -> Weight {
+                // Migration intended only for the specific runtime update from version 5_002
+                // (which uses pallet_staking::UseNominatorsAndValidatorsMap as VoterList)
+                if crate::System::last_runtime_upgrade_spec_version() == 5_002 {
+                    let migrated =
+                        <Runtime as pallet_staking::Config>::VoterList::unsafe_regenerate(
+                            pallet_staking::Validators::<Runtime>::iter()
+                                .map(|(id, _)| id)
+                                .chain(
+                                    pallet_staking::Nominators::<Runtime>::iter().map(|(id, _)| id),
+                                ),
+                            crate::Staking::weight_of_fn(),
+                        );
+
+                    log::info!(
+                        "👜 completed staking migration to bags list with {} voters migrated",
+                        migrated,
+                    );
+
+                    BlockWeights::get().max_block
+                } else {
+                    <Runtime as frame_system::Config>::DbWeight::get().reads(1)
+                }
```

Comment:
I prefer the _guard_ approach instead... But feel free to leve it as it is.

Reply:
Thanks for the tip, I'll look into that for future reference. Anyways, this is what Polkadot did for the equivalent migration and it looks reasonable in this case, I'd leave it as it is this time.